tests/driver.sh

```diff
@@ -1272,4 +1272,261 @@ int main()
+    if (func(b) || func(c))
+        printf("%d %d\n", b, c);
+
+    if (func(d + 100) || func(b))
+        printf("%d %d\n", b, c);
+    else
+        printf("get 0\n");
+
+
+    return 0;
+}
+EOF
+
+# Logical-or: More complex use cases
+ans="a && b && c && d = 0
+a || b && c && d = 1
+a && b || c && d = 1
+a && b && c || d = 1
+a || b || c && d = 1
+a || b && c || d = 1
+a && b || c || d = 1
+a || b || c || d = 0
+(a || b) && c && d = 0
+a && (b || c) && d = 1
+a && b && (c || d) = 1
+(a || b || c) && d = 1
+(a || b) && (c || d) = 1
+a && (b || c || d) = 1
+a * 0 && (b || c || d) = 0
+a * 2 && (b || c || d) = 1
+a && (b * 0 || c || d * 0) = 0
+func(10): 10 > 0
+func(20): 20 > 0
+func(0): 0 <= 0
+func(a) && func(b) && func(c) = 0
+func(10): 10 > 0
+func(a) - a && func(b) && func(c) = 0
+func(10): 10 > 0
+func(a) && func(b) && func(c) + 1 = 0
+func(0): 0 <= 0
+func(10): 10 > 0
+func(c) || func(a) && func(d) = 1
+func(0): 0 <= 0
+func(0): 0 <= 0
+func(-100): -100 <= 0
+func(c) || func(c) && func(d) = 1
+func(0): 0 <= 0
+func(0): 0 <= 0
+(func(c) || func(c)) && func(d + 100) = 0
+func(0): 0 <= 0
+func(10): 10 > 0
+func(-100): -100 <= 0
+func(c) || func(a) && func(d) = 1
+func(10): 10 > 0
+func(-100): -100 <= 0
+func(a) && (func(d) || func(c)) = 1
+func(10): 10 > 0
+func(-100): -100 <= 0
+func(a) && (func(d) || func(c)) = 1"
+
+try_output 0 "$ans" << EOF
+int func(int x)
+{
+    if (x > 0)
+        printf("func(%d): %d > 0\n", x, x);
+    else
+	printf("func(%d): %d <= 0\n", x, x);
+    return x;
+}
+
+int main()
+{
+    int a = 10, b = 20, c = 0, d = -100;
+    printf("a && b && c && d = %d\n", a && b && c && d);
+    printf("a || b && c && d = %d\n", a || b && c && d);
+    printf("a && b || c && d = %d\n", a && b || c && d);
+    printf("a && b && c || d = %d\n", a && b && c || d);
+    printf("a || b || c && d = %d\n", a || b || c && d);
+    printf("a || b && c || d = %d\n", a || b && c || d);
+    printf("a && b || c || d = %d\n", a && b || c || d);
+    printf("a || b || c || d = %d\n", a && b && c && d);
+
+    printf("(a || b) && c && d = %d\n", (a || b) && c && d);
+    printf("a && (b || c) && d = %d\n", a && (b || c) && d);
+    printf("a && b && (c || d) = %d\n", a && b && (c || d));
+    printf("(a || b || c) && d = %d\n", (a || b || c) && d);
+    printf("(a || b) && (c || d) = %d\n", (a || b) && (c || d));
+    printf("a && (b || c || d) = %d\n", a && (b || c || d));
+    printf("a * 0 && (b || c || d) = %d\n", a * 0 && (b || c || d));
+    printf("a * 2 && (b || c || d) = %d\n", a * 2 && (b || c || d));
+    printf("a && (b * 0 || c || d * 0) = %d\n", a && (b * 0 || c || d * 0));
+
+    printf("func(a) && func(b) && func(c) = %d\n", func(a) && func(b) && func(c));
+    printf("func(a) - a && func(b) && func(c) = %d\n", func(a) - a && func(b) && func(c));
+    printf("func(a) && func(b) && func(c) + 1 = %d\n", func(a) - a && func(b) && func(c) + 1);
+    printf("func(c) || func(a) && func(d) = %d\n", func(c) || func(a) || func(d));
```

Comment:
Is the difference between the printed statement and the function evaluation intentional?